Firestore/core/test/firebase/firestore/testutil/async_testing.h

```diff
@@ -46,6 +52,84 @@ std::unique_ptr<util::Executor> ExecutorForTesting(const char* name);
  */
 std::shared_ptr<util::AsyncQueue> AsyncQueueForTesting();
 
+constexpr auto kTimeout = std::chrono::seconds(5);
+
+/**
+ * An expected outcome of an asynchronous test.
+ */
+class Expectation {
+ public:
+  Expectation();
+
+  /** Marks this expectation as fulfilled. */
+  void Fulfill();
+
+  /**
+   * Returns a callback function, that when invoked, fullfills the expectation.
+   *
+   * The returned function has a lifetime that's independent of the Expectation
+   * that created it.
+   */
+  std::function<void()> AsCallback() const;
+
+  /**
+   * Returns a `future` that represents the completion of this Expectation.
+   */
+  std::future<void> get_future() const;
+
+ private:
+  std::shared_ptr<std::promise<void>> promise_;
+};
+
+/**
+ * A mixin that supplies utilities for safely writing asynchronous tests.
+ */
+class AsyncTest {
+ public:
+  AsyncTest()
+      : trace_("Test case name",
+               1,
+               testing::Message() << testing::UnitTest::GetInstance()
+                                         ->current_test_info()
+                                         ->name()) {
+  }
+
+  std::future<void> Async(std::function<void()> action);
+
+  /**
+   * Waits for the future to become ready.
+   *
+   * Fails the current test if the timeout occurs.
+   */
+  void Await(const std::future<void>& future,
+             std::chrono::milliseconds timeout = kTimeout);
+
+  /**
+   * Waits for the promise to become ready. The future associated with the
+   * promise is consumed via `promise.get_future()`.
+   *
+   * Fails the current test if the timeout occurs.
+   */
+  void Await(std::promise<void>& promise,  // NOLINT(runtime/references)
```

Comment:
This is, strictly speaking, against style guide, though I'm certainly not complaining -- feel free to leave as is.